cmd/kubeadm/app/util/pkiutil/pki_helpers.go

```diff
@@ -556,32 +625,39 @@ func NewPrivateKey(keyType x509.PublicKeyAlgorithm) (crypto.Signer, error) {
 }
 
 // NewSignedCert creates a signed certificate using the given CA certificate and key
-func NewSignedCert(cfg *CertConfig, key crypto.Signer, caCert *x509.Certificate, caKey crypto.Signer) (*x509.Certificate, error) {
+func NewSignedCert(cfg *CertConfig, key crypto.Signer, caCert *x509.Certificate, caKey crypto.Signer, isCA bool) (*x509.Certificate, error) {
 	serial, err := cryptorand.Int(cryptorand.Reader, new(big.Int).SetInt64(math.MaxInt64))
 	if err != nil {
 		return nil, err
 	}
 	if len(cfg.CommonName) == 0 {
 		return nil, errors.New("must specify a CommonName")
 	}
-	if len(cfg.Usages) == 0 {
+	if !isCA && len(cfg.Usages) == 0 {
 		return nil, errors.New("must specify at least one ExtKeyUsage")
 	}
 
+	keyUsage := x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature
+	if isCA {
+		keyUsage |= x509.KeyUsageCertSign
+	}
+
 	RemoveDuplicateAltNames(&cfg.AltNames)
 
 	certTmpl := x509.Certificate{
 		Subject: pkix.Name{
 			CommonName:   cfg.CommonName,
 			Organization: cfg.Organization,
 		},
-		DNSNames:     cfg.AltNames.DNSNames,
-		IPAddresses:  cfg.AltNames.IPs,
-		SerialNumber: serial,
-		NotBefore:    caCert.NotBefore,
-		NotAfter:     time.Now().Add(kubeadmconstants.CertificateValidity).UTC(),
-		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
-		ExtKeyUsage:  cfg.Usages,
+		DNSNames:              cfg.AltNames.DNSNames,
+		IPAddresses:           cfg.AltNames.IPs,
+		SerialNumber:          serial,
+		NotBefore:             caCert.NotBefore,
+		NotAfter:              time.Now().Add(kubeadmconstants.CertificateValidity).UTC(),
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           cfg.Usages,
+		BasicConstraintsValid: isCA,
+		IsCA:                  isCA,
```

Comment:
https://golang.org/pkg/crypto/x509/

>     // BasicConstraintsValid indicates whether IsCA, MaxPathLen,
>    // and MaxPathLenZero are valid.

the documentation is a bit vague on what "valid" means, but should we assume `BasicConstraintsValid` should be always bound to `isCA`?

e.g. here it's not:
https://github.com/kubernetes/kubernetes/blob/dbfc3aa877b5beda2ecdfb38911c311b250a4973/pkg/controller/certificates/authority/policies.go#L65-L66

can we pass `true` always to it and an explicit value to `MaxPathLen` and `MaxPathLenZero=false`?

i'm assuming this is where it's checked:
https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L721-L726

Reply:
I just did it this way to make sure I didn't make any changes to the existing code which signs leaf certificates. My interpretation of the docs is that it would probably work to set `BasicConstraintsValid` to `true`. We shouldn't need to explicitly set `MaxPathLen` and `MaxPathLenZero` since the default values of 0 and false respectively are "interpreted as `MaxPathLen` not being set" according the docs.

Reply:
my vote goes to set it always to `true` even for non-CA certificates.
`BasicConstraintsValid` seems like a Golang wrapper over the Basic Constraints extension, to conform checks in crypto/x509/verify.go and it doesn't seem like a breaking change.